### ISSUE TITLE
Jackson not working on Serialization.BEAN annotated class in release build(proguard)

### ISSUE THREAD

**Issue by MatFl** [johncarl81/parceler#271]
I seem to have stumbled upon a very strange issue. When my model class is annotated with `@Parcel(Parcel.Serialization.BEAN)`, I can no longer use Jackson to convert it to JSON. It will always throw a `java.lang.NoSuchFieldError`, but without a name of the field.

If I use the plain @Parcel annotation and @ParcelConstructor, or don't use proguard at all, it is working. Here is an example class:

```
@Parcel(Parcel.Serialization.BEAN)
public class TestItem {
    private boolean selected;

    public TestItem() {
        // bean constructor
    }

    public TestItem(final boolean selected) {
        this.selected = selected;
    }

    public boolean isSelected() {
        return selected;
    }

    public void setSelected(final boolean selected) {
        this.selected = selected;
    }
}
```

I have added my own classes, the recommended parceler and jackson configuration for proguard.

So I am not sure what exactly is causing the problem. I skipped over some of the processor code but could not find anything that Parceler would change the original class.

Any ideas what could cause this?

**Comment by johncarl81** [comment]
You're right, Parceler will not (can not) chat the original class.  Can you share a working example of the issue?

**Comment by MatFl** [comment]
I can't seem to replicate this in an example project and I can't share my actual project. So I have to try some more things first.

However I got the same error with another class, that also has the BEAN annotation. I dug a bit deeper and got the root cause of the error:

> java.lang.NoSuchFieldException: BEAN
> at java.lang.Class.getDeclaredField(Class.java:886)
> at libcore.reflect.AnnotationAccess.decodeValue(AnnotationAccess.java:685)
> at libcore.reflect.AnnotationAccess.toAnnotationInstance(AnnotationAccess.java:663)
> at libcore.reflect.AnnotationAccess.annotationSetToAnnotations(AnnotationAccess.java:633)
> at libcore.reflect.AnnotationAccess.getDeclaredAnnotations(AnnotationAccess.java:161)
> at java.lang.Class.getDeclaredAnnotations(Class.java:861)
> at com.fasterxml.jackson.databind.introspect.AnnotatedClass.resolveClassAnnotations(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.AnnotatedClass.getAnnotation(Unknown Source)
> at com.fasterxml.jackson.databind.AnnotationIntrospector._findAnnotation(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.JacksonAnnotationIntrospector.findAutoDetectVisibility(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.<init>(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.constructPropertyCollector(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.collectProperties(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.forSerialization(Unknown Source)
> at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.forSerialization(Unknown Source)
> at com.fasterxml.jackson.databind.SerializationConfig.introspect(Unknown Source)
> at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(Unknown Source)
> at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(Unknown Source)
> at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(Unknown Source)
> at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(Unknown Source)
> at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(Unknown Source)
> at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(Unknown Source)
> at com.fasterxml.jackson.databind.ObjectMapper._configAndWriteValue(Unknown Source)
> at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(Unknown Source)
> at xyz.MyTest.shouldConvertToJson(MyTest.java:41)
> at java.lang.reflect.Method.invoke(Native Method)
> at java.lang.reflect.Method.invoke(Method.java:372)
> at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
> at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
> at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
> at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
> at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
> at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
> at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
> at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
> at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
> at org.junit.runners.Suite.runChild(Suite.java:128)
> at org.junit.runners.Suite.runChild(Suite.java:27)
> at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
> at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
> at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
> at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
> at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
> at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
> at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
> at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
> at android.support.test.internal.r

So the BEAN annotation seems to be part of this. Somehow the annotation seems to be handled as field?
Additionally I get the same error, when trying to mock this class with mockito.

**Comment by johncarl81** [comment]
Very strange...

I tried your example out with the following Mockito mock without error:
```java
public class TestItemTest {

    @Test
    public void mockTestItem() {
        TestItem item = mock(TestItem.class);
    }
}
```
I'm using mockito 1.10.19.  Does this fail for you?

**Comment by MatFl** [comment]
Hi,

So I got it working, just as I finished the example project. In the example I also noticed that a class with the plain @Parcel annotation could not be mocked (jackson was working though).

I had to change the proguard config a bit:

```
-keep interface org.parceler.Parcel
-keep enum org.parceler.Parcel$Serialization { *; }  # added
-keep @org.parceler.Parcel class * { *; }
-keep class **$$Parcelable { *; }
```

Now mockito and jackson are working correct with the parceled class. I think this should be added to the readme.

I am using Mockito 2.7. They hava a mockito-android module for a few versions now, so you don't have to use the old version and dexmaker anymore.

If you are still interested in the example project I can upload it later.
